### PR TITLE
Add superfluid connector

### PR DESCRIPTION
**Pull request by CoffeeFam84**
Add superfluid connector to arbitrum, avalanche, optimism and polygon.

### CODE REVIEW

File: contracts/arbitrum/connectors/superfluid/events.sol [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	 * @param token Super token address
+	 * @param sender Flow sender address
+	 * @param receiver Flow recipient address
+	 * @param flowRate Flow rate in amount per second for this flow
+	 * @param totalSenderFlowRate Total flow rate in amount per second for the sender
+	 * @param totalReceiverFlowRate Total flow rate in amount per second for the receiver
+	 * @param userData The user provided data
+	 *
+	 */
+	event FlowUpdated(
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/events.sol]: Every event should start with `Log...`
```suggestion
	event LogFlowUpdated(

```

File: contracts/arbitrum/connectors/superfluid/events.sol [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		ISuperfluidToken indexed token,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/events.sol]: Can we emit it as address?

File: contracts/arbitrum/connectors/superfluid/helpers.sol [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+abstract contract Helpers {
+	using CFAv1Library for CFAv1Library.InitData;
+	using IDAv1Library for IDAv1Library.InitData;
+
+	ISuperfluid host = ISuperfluid(0xCf8Acb4eF033efF16E8080aed4c7D5B9285D2192);
+	IInstantDistributionAgreementV1 ida =
+		IInstantDistributionAgreementV1(
+			0x2319C7e07EB063340D2a0E36709B0D65fda75986
+		);
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/helpers.sol]: should be an internal constant.
```suggestion
	ISuperfluid internal constant host = ISuperfluid(0xCf8Acb4eF033efF16E8080aed4c7D5B9285D2192);
	IInstantDistributionAgreementV1 internal constant ida =
		IInstantDistributionAgreementV1(
			0x2319C7e07EB063340D2a0E36709B0D65fda75986
		);

```

File: contracts/arbitrum/connectors/superfluid/helpers.sol [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+	//initialize InitData struct, and set equal to cfaV1
+	CFAv1Library.InitData public cfaV1 =
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/helpers.sol]:
```suggestion
	CFAv1Library.InitData internal constant cfaV1 =

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+		ISuperToken superToken,
+		uint256 amount,
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		IERC20(superToken.getUnderlyingToken()).approve(
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:37]: We can get the amount using getId as well, and this amount should be passed to approve func
```suggestion
uint256 amt_ = getUint(getId, amount);
		IERC20(superToken.getUnderlyingToken()).approve(

```

**Comment by pradyuman-verma** [comment]
@CoffeeFam84 
- The use of `getId` and `setId` is missing in functions.
- Variables should be `internal constant.`
- You also need to return the `_eventName, _eventParam` for each function.

You can check aave v3 connector, for example - https://github.com/Instadapp/dsa-connectors/blob/main/contracts/polygon/connectors/aave/v3/main.sol

**Comment by CoffeeFam84** [comment]
For `getId` and `setId`, we don't need them in functions, in my point of view.
And for emitting events, every function in superfluid are emitting events already. so do we need to emit them again in instadapp?

**Comment by pradyuman-verma** [comment]
> 

Yes we do, these events help us do to tracking related to DSAs

**Comment by pradyuman-verma** [comment]
> @pradyuman-verma For `getId` and `setId`, we don't need them in functions, in my point of view. And for emitting events, every function in superfluid are emitting events already. so do we need to emit them again in instadapp?

@CoffeeFam84 setId and getId are used to communicate between simultaneous spells, that's why those are necessary

File: contracts/arbitrum/connectors/superfluid/helpers.sol [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+						keccak256(
+							"org.superfluid-finance.agreements.ConstantFlowAgreement.v1"
+						)
+					)
+				)
+			)
+		);
+
+	// declare `_idaLib` of type InitData and assign it the host and ida addresses
+	IDAv1Library.InitData internal _idav1Lib = IDAv1Library.InitData(host, ida);
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/helpers.sol:37]: Should be constant as well

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 7]
```diff
@@ -0,0 +1,768 @@
...
+	 * @notice Delete an existing stream.
+	 * @param sender The sender of the stream
+	 * @param receiver The receiver of the stream
+	 * @param superToken The super token which will be streamed to the receiver
+	 * @param userData Optional bytes value for passing in arbitrary data along with the transaction.
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of token.
+	 */
+	function deleteFlow(
+		address sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:174]: We don't need sender, the sender will be `address(this)`

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 8]
```diff
@@ -0,0 +1,768 @@
...
+		ISuperToken superToken,
+		bytes calldata userData,
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		CFAv1Library.deleteFlow(cfaV1, sender, receiver, superToken, userData);
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:185]:
```suggestion
		CFAv1Library.deleteFlow(cfaV1, address(this), receiver, superToken, userData);

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 9]
```diff
@@ -0,0 +1,768 @@
...
+	 * @param sender The sender on the flow
+	 * @param receiver The receiver of the flow
+	 * @param flowRate Number of tokens sent to the receiver per second (denominated in wei)
+	 * @param userData Optional bytes value for passing in arbitrary data along with the transaction.
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of token.
+	 */
+	function createFlowByOperator(
+		ISuperToken superToken,
+		address sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:322]:
```suggestion
		

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 10]
```diff
@@ -0,0 +1,768 @@
...
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		CFAv1Library.createFlowByOperator(
+			cfaV1,
+			sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:335]:
```suggestion
			addresss(this),

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 11]
```diff
@@ -0,0 +1,768 @@
...
+	 * @param sender The sender on the flow
+	 * @param receiver The receiver of the flow
+	 * @param flowRate Number of tokens sent to the receiver per second (denominated in wei)
+	 * @param userData Optional bytes value for passing in arbitrary data along with the transaction.
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of token.
+	 */
+	function updateFlowByOperator(
+		ISuperToken superToken,
+		address sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:367]:
```suggestion
		

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 12]
```diff
@@ -0,0 +1,768 @@
...
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		CFAv1Library.updateFlowByOperator(
+			cfaV1,
+			sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:380]:
```suggestion
		address(this),

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 13]
```diff
@@ -0,0 +1,768 @@
...
+	 * @param superToken The super token on which the flow is being deleted by operator
+	 * @param sender The sender on the flow
+	 * @param receiver The receiver of the flow
+	 * @param userData Optional bytes value for passing in arbitrary data along with the transaction.
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of token.
+	 */
+	function deleteFlowByOperator(
+		ISuperToken superToken,
+		address sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:411]:
```suggestion
		

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 14]
```diff
@@ -0,0 +1,768 @@
...
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		CFAv1Library.deleteFlowByOperator(
+			cfaV1,
+			sender,
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:423]:
```suggestion
			address(this),

```

File: contracts/arbitrum/connectors/superfluid/main.sol [hunk 15]
```diff
@@ -0,0 +1,768 @@
...
+		bytes calldata userData,
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		IDAv1Library.distribute(_idav1Lib, superToken, id, amount, userData);
+
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/superfluid/main.sol:540]: We can use getId here to exact amount.

**Comment by thrilok209** [comment]
Since no activity from the past 1 month from the contributor, closing this pr for now. 

Later it can be re-open suggested changes are made. 

Thank you

**Pull request closed without merging**